src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -3574,68 +3574,168 @@ internal Binder GetBinder(CSharpSyntaxNode syntaxNode)
             memberNames.AddRange(membersByName.Keys);
 
             //key and value will be the same object
-            var methodsBySignature = new Dictionary<MethodSymbol, SourceMemberMethodSymbol>(MemberSignatureComparer.PartialMethodsComparer);
+            var membersBySignature = new Dictionary<Symbol, Symbol>(MemberSignatureComparer.PartialMethodsComparer);
 
             foreach (var name in memberNames)
             {
-                methodsBySignature.Clear();
+                membersBySignature.Clear();
                 foreach (var symbol in membersByName[name])
                 {
-                    var method = symbol as SourceMemberMethodSymbol;
-                    if (method is null || !method.IsPartial)
+                    if (!symbol.IsPartialMember())
                     {
-                        continue; // only partial methods need to be merged
+                        continue;
                     }
 
-                    if (methodsBySignature.TryGetValue(method, out var prev))
+                    if (!membersBySignature.TryGetValue(symbol, out var prev))
                     {
-                        var prevPart = (SourceOrdinaryMethodSymbol)prev;
-                        var methodPart = (SourceOrdinaryMethodSymbol)method;
+                        membersBySignature.Add(symbol, symbol);
+                        continue;
+                    }
 
-                        if (methodPart.IsPartialImplementation &&
-                            (prevPart.IsPartialImplementation || (prevPart.OtherPartOfPartial is MethodSymbol otherImplementation && (object)otherImplementation != methodPart)))
-                        {
-                            // A partial method may not have multiple implementing declarations
-                            diagnostics.Add(ErrorCode.ERR_PartialMethodOnlyOneActual, methodPart.GetFirstLocation());
-                        }
-                        else if (methodPart.IsPartialDefinition &&
-                                 (prevPart.IsPartialDefinition || (prevPart.OtherPartOfPartial is MethodSymbol otherDefinition && (object)otherDefinition != methodPart)))
-                        {
-                            // A partial method may not have multiple defining declarations
-                            diagnostics.Add(ErrorCode.ERR_PartialMethodOnlyOneLatent, methodPart.GetFirstLocation());
-                        }
-                        else
-                        {
-                            if ((object)membersByName == _lazyEarlyAttributeDecodingMembersDictionary)
+                    Debug.Assert(symbol.GetType() == prev.GetType());
```

Comment:
nit: this assertion seems redundant because we throw in the `default` case below

Reply:
My intent here was to enforce that the exact same subtype is in use (e.g. if one part is a SourceOrdinaryMethodSymbolComplex, then so should the other part). But, I can't actually think of anything important which is preserved by ensuring this, so I'll remove the assertion.